namer/marathon/src/main/scala/io/buoyant/namer/marathon/BasicAuthenticator.scala

```diff
@@ -0,0 +1,48 @@
+package io.buoyant.namer.marathon
+
+import com.twitter.finagle.{Failure, http}
+import com.twitter.util.{Future, Return, Throw, Time}
+import io.buoyant.marathon.v2.Api
+import java.util.concurrent.atomic.AtomicReference
+import scala.annotation.tailrec
+
+object BasicAuthenticator {
+
+  private[marathon] case class UnauthorizedResponse(rsp: http.Response) extends Throwable
+  private[this] val closedException = Failure("closed", Failure.Interrupted)
+  private[this] val closedExceptionF = Future.exception(closedException)
+
+  class Authenticated(client: Api.Client, http_auth_token: String) extends Api.Client {
+    private[this] sealed trait State
+    private[this] object Closed extends State
+    private[this] object Init extends State
+    private[this] val state = new AtomicReference[State](Init)
```

Comment:
i think we can do away with most of the state management, as most of that was motivated by the dc/os auth module hold the state of the authentication flow.

```scala
class BasicAuthenticatorFilter(http_auth_token: String) extends Filter[http.Request, http.Response] {
  def apply(req: http.Request, svc: Service[http.Request, http.Response]): Future[http.Response] = {
    req.headerMap.set("Authorization", s"Basic $token")
    svc(req)
  }
}
```